File: backend/app/schemas/booking.py
"""
Booking Schemas
Pydantic models for booking-related request/response validation.
"""

from pydantic import BaseModel, Field, field_validator
from typing import Optional, List
from uuid import UUID
from datetime import datetime, date
from decimal import Decimal
from ..models.booking import BookingStatus
from .service import ServiceResponse
from .user import UserResponse


class BookingServiceInfo(BaseModel):
    """Schema for service info in a booking."""
    service_id: UUID
    service_name: str
    service_price: Decimal
    
    class Config:
        from_attributes = True


class BookingCreate(BaseModel):
    """Schema for creating a new booking."""
    service_ids: List[UUID] = Field(..., min_length=1)
    booking_date: date
    notes: Optional[str] = Field(None, max_length=500)
    
    @field_validator('booking_date')
    @classmethod
    def validate_booking_date(cls, v):
        if v < date.today():
            raise ValueError('Booking date cannot be in the past')
        return v


class BookingUpdate(BaseModel):
    """Schema for updating a booking."""
    booking_date: Optional[date] = None
    notes: Optional[str] = Field(None, max_length=500)
    
    @field_validator('booking_date')
    @classmethod
    def validate_booking_date(cls, v):
        if v and v < date.today():
            raise ValueError('Booking date cannot be in the past')
        return v


class BookingStatusUpdate(BaseModel):
    """Schema for updating booking status (Owner only)."""
    status: BookingStatus
    
    @field_validator('status')
    @classmethod
    def validate_status(cls, v):
        # Define valid status transitions
        valid_statuses = [
            BookingStatus.CONFIRMED,
            BookingStatus.IN_PROGRESS,
            BookingStatus.READY_FOR_DELIVERY,
            BookingStatus.COMPLETED,
            BookingStatus.CANCELLED
        ]
        if v not in valid_statuses:
            raise ValueError(f'Invalid status. Must be one of: {[s.value for s in valid_statuses]}')
        return v


class BookingResponse(BaseModel):
    """Schema for booking response."""
    id: UUID
    customer: UserResponse
    booking_date: date
    status: BookingStatus
    status_display: str
    total_price: Decimal
    notes: Optional[str]
    services: List[BookingServiceInfo]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class BookingListResponse(BaseModel):
    """Schema for list of bookings response."""
    bookings: List[BookingResponse]
    total: int
    page: int
    page_size: int


class BookingCustomerView(BaseModel):
    """Schema for customer's view of their booking."""
    id: UUID
    booking_date: date
    status: BookingStatus
    status_display: str
    total_price: Decimal
    notes: Optional[str]
    services: List[BookingServiceInfo]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

File: frontend/lib/models/booking.dart
import 'user.dart';

/// Booking status enum
enum BookingStatus {
  pending,
  confirmed,
  inProgress,
  readyForDelivery,
  completed,
  cancelled;

  static BookingStatus fromString(String value) {
    switch (value) {
      case 'pending':
        return BookingStatus.pending;
      case 'confirmed':
        return BookingStatus.confirmed;
      case 'in_progress':
        return BookingStatus.inProgress;
      case 'ready_for_delivery':
        return BookingStatus.readyForDelivery;
      case 'completed':
        return BookingStatus.completed;
      case 'cancelled':
        return BookingStatus.cancelled;
      default:
        return BookingStatus.pending;
    }
  }

  String get apiValue {
    switch (this) {
      case BookingStatus.pending:
        return 'pending';
      case BookingStatus.confirmed:
        return 'confirmed';
      case BookingStatus.inProgress:
        return 'in_progress';
      case BookingStatus.readyForDelivery:
        return 'ready_for_delivery';
      case BookingStatus.completed:
        return 'completed';
      case BookingStatus.cancelled:
        return 'cancelled';
    }
  }

  String get displayName {
    switch (this) {
      case BookingStatus.pending:
        return 'Pending';
      case BookingStatus.confirmed:
        return 'Confirmed';
      case BookingStatus.inProgress:
        return 'In Progress';
      case BookingStatus.readyForDelivery:
        return 'Ready for Delivery';
      case BookingStatus.completed:
        return 'Completed';
      case BookingStatus.cancelled:
        return 'Cancelled';
    }
  }
}

/// Booking service info
class BookingServiceInfo {
  final String serviceId;
  final String serviceName;
  final double servicePrice;

  BookingServiceInfo({
    required this.serviceId,
    required this.serviceName,
    required this.servicePrice,
  });

  factory BookingServiceInfo.fromJson(Map<String, dynamic> json) {
    return BookingServiceInfo(
      serviceId: json['service_id'],
      serviceName: json['service_name'],
      servicePrice: double.parse(json['service_price'].toString()),
    );
  }
}

/// Booking model
class Booking {
  final String id;
  final User customer;
  final DateTime bookingDate;
  final BookingStatus status;
  final String statusDisplay;
  final double totalPrice;
  final String? notes;
  final List<BookingServiceInfo> services;
  final DateTime createdAt;
  final DateTime updatedAt;

  Booking({
    required this.id,
    required this.customer,
    required this.bookingDate,
    required this.status,
    required this.statusDisplay,
    required this.totalPrice,
    this.notes,
    required this.services,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'],
      customer: User.fromJson(json['customer']),
      bookingDate: DateTime.parse(json['booking_date']),
      status: BookingStatus.fromString(json['status']),
      statusDisplay: json['status_display'] ?? '',
      totalPrice: double.parse(json['total_price'].toString()),
      notes: json['notes'],
      services: (json['services'] as List)
          .map((s) => BookingServiceInfo.fromJson(s))
          .toList(),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }
}

/// Booking list response
class BookingListResponse {
  final List<Booking> bookings;
  final int total;
  final int page;
  final int pageSize;

  BookingListResponse({
    required this.bookings,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  factory BookingListResponse.fromJson(Map<String, dynamic> json) {
    return BookingListResponse(
      bookings: (json['bookings'] as List)
          .map((b) => Booking.fromJson(b))
          .toList(),
      total: json['total'],
      page: json['page'],
      pageSize: json['page_size'],
    );
  }
}
